Add NodePort support for Antrea Proxy on Linux

weiqiangt opened the pull request:
Note: To verify this PR, the NodePort support is enabled. The feature gate should be disabled before merging this PR.
Resolves #1463.
Signed-off-by: Weiqiang Tang <weiqiangt@vmware.com>

weiqiangt commented:
/test-all

tnqn commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -116,3 +119,13 @@ featureGates:
 #nplPortRange: 40000-41000
+
+# The virtual IP for NodePort Service support. It must be a link-local IP otherwise the Agents will report error.
+#nodePortVirtualIP: 169.254.169.110
```
Have you tried policy route approach? I did an experiment and it worked as we discussed. 
If NAT is not likely to be the final solution, better not to add this argument in case compatbility issue in future. We can just use a reserved IP in alpha phase I think.

jianjuns replied:
Why we care it must be link-local?

weiqiangt replied:
@tnqn, I tried the policy route way.
If the destination is a loopback address we still need to do a DNAT. Considering this is an ALPHA feature for now, we can discuss it more in the future.

weiqiangt replied:
@jianjuns, do you think we should use any forwardable address here?

jianjuns replied:
I want to understand the current design first. Why IPv4 must be loopback, and IPv6 must not be?

jianjuns replied:
Got more understanding after reading the design proposal. I think no need to restrict the IP must be link-local. We can recommend link-local though.

weiqiangt replied:
Thanks @jianjuns, updated.

tnqn replied:
I mean there are possibilities we don't need to do DNAT in future like the two approaches we have discussed. Even if we still need it for loopback address ( maybe we don't need it for this case as well), I think hardcoding a reserved IP is enough at this stage.
I just want to avoid introducing implementation specific configurations when it's not stable so we don't have the complexity of upgrade (If user set it and we delete it in future, the process will crash). 
And it doesn't seem that user can get what the configuration means and how to use it with the current comments.

weiqiangt replied:
@tnqn, thanks for the explanation. I removed these two options and made them constant for now.

tnqn reviewed (commented):
I haven't finished review. But I did some test based on the patch and found an issue:
When scaling down backend pods, the members in openflow group doesn't change, leading to connectivity issue.

jianjuns commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -120,3 +123,13 @@ featureGates:
 # Defaults to "". It must be a host string, a host:port pair, or a URL to the base of the apiserver.
 #kubeAPIServerOverride: ""
+
+# The virtual IP for NodePort Service support. It must be a link-local IP otherwise the Agents will report error.
```
Probably change Agents to antrea-agent to be consistent with other references in this file.

weiqiangt replied:
Done.

jianjuns commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -120,3 +123,13 @@ featureGates:
+# The virtual IP for NodePort Service support. It must be a link-local IP otherwise the Agents will report error.
+#nodePortVirtualIP: 169.254.169.110
+
+# The virtual IPv6 for NodePort Service support. It must not be a link-local IP otherwise the Agents will report error.
```
Why it cannot be link-local?

weiqiangt replied:
The IPv6 forwarding is stricter than IPv4, and link-local IP is not forwardable.

jianjuns replied:
Ok.

jianjuns commented on cmd/antrea-agent/agent.go:
```diff
@@ -180,17 +184,27 @@ func run(o *Options) error {
 
 	var proxier k8sproxy.Provider
 	if features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
+		var nodePortAddresses []*net.IPNet
```
I feel we should have a separate feature gate to control NodePort by AntreaProxy or not.

weiqiangt replied:
Yes, I added a new feature gate `AntreaProxyNodePort` to control it. Also, I updated the workflow here. The parsing of NodeAddress will go only if both `AntreaProxyNodePort` and `AntreaProxy` are enabled.

jianjuns reviewed (commented):
Do you have a design doc for the reviewers to understand the design?

weiqiangt commented:
> Do you have a design doc for the reviewers to understand the design?

I attached the issue which includes the design.

weiqiangt commented:
> I haven't finished review. But I did some test based on the patch and found an issue:
> When scaling down backend pods, the members in openflow group doesn't change, leading to connectivity issue.

Thanks @tnqn. I have fixed this issue and added checks in e2e.

jianjuns commented:
> I attached the issue which includes the design.
Thanks! It is much clear to me now after reading the design description in the issue.

weiqiangt commented:
/test-all

weiqiangt commented:
/test-all

weiqiangt commented:
Windows conformance failed due to infrastructure failure, re-trigger it.
/test-windows-conformance

weiqiangt commented:
/test-all

weiqiangt commented:
/jenkins-ipv6-ds-conformance
jenkins-ipv6-only-conformance

jianjuns commented:
I do not mean to ask for a design change for this PR, but have we considered usign TC to redirect traffic (https://man7.org/linux/man-pages/man8/tc-mirred.8.html)? Could it be faster or slower than iptables?

weiqiangt commented:
/test-all

Jenkins tests failed on some unrelated cases, re-trigger tests to have a double-check.
```
[k8s.io] InitContainer [NodeConformance] should not start app containers and fail the pod if init containers fail on a RestartNever pod [Conformance]
[k8s.io] InitContainer [NodeConformance] should invoke init containers on a RestartAlways pod [Conformance]
[k8s.io] Security Context When creating a pod with readOnlyRootFilesystem should run the container with writable rootfs when readOnlyRootFilesystem=false [NodeConformance] [Conformance]
[k8s.io] Probing container should be restarted with a exec "cat /tmp/health" liveness probe [NodeConformance] [Conformance]
[k8s.io] Kubelet when scheduling a busybox Pod with hostAliases should write entries to /etc/hosts [LinuxOnly] [NodeConformance] [Conformance]
```

weiqiangt commented:
/test-ipv6-ds-conformance
/test-ipv6-only-conformance

weiqiangt commented:
/test-windows-conformance

jianjuns commented on build/yamls/antrea.yml:
```diff
@@ -1283,6 +1283,9 @@ data:
     #  AntreaProxy: true
 
+    # Enable NodePort Service support in AntreaProxy in antrea-agent.
+      AntreaProxyNodePort: true
```
I would suggest not to enable it by default. Even we enable it, it does not provide any value, until we have a solution to remove kube-proxy.
@tnqn @antoninbas

antoninbas replied:
Agreed

tnqn replied:
This is the plan. I think Weiqiang only enables it for testing. See PR description "To verify this PR, the NodePort support is enabled. The feature gate should be disabled before merging this PR."

weiqiangt replied:
Yes, it is enabled only for testing. I will disable it when we're ready to merge.

jianjuns commented on pkg/agent/route/route_linux.go:
```diff
@@ -657,3 +794,145 @@ func (c *Client) UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error
 	return nil
 }
+
+func (c *Client) ReconcileNodePort(nodeIPs []net.IP, svcEntries []*proxytypes.ServiceInfo) error {
```
I did not see this func is called, but might be my mistake?

jianjuns commented on pkg/agent/route/interfaces.go:
```diff
@@ -38,6 +39,20 @@ type Interface interface {
 	DeleteRoutes(podCIDR *net.IPNet) error
 
+	// AddRoutes should add the route to the NodePort virtual IP.
+	AddNodePortRoute(isIPv6 bool) error
```
What you think adding a separate interface for AddNodePortRoute, AddNodePort, and DeleteNodePort, which can have different implementations in Linux (by Route Client) and Windows (by Openflow)? Or maybe on Windows they are not needed at all, as the existing Openflow Client interfaces can already cover these functions?
